src/logging.c

```diff
@@ -805,30 +853,102 @@ int rcutils_logging_get_logger_effective_level(const char * name)
   if (NULL == name) {
     return -1;
   }
-  size_t substring_length = strlen(name);
-  while (true) {
-    int severity = rcutils_logging_get_logger_leveln(name, substring_length);
-    if (-1 == severity) {
-      RCUTILS_SAFE_FWRITE_TO_STDERR_WITH_FORMAT_STRING(
-        "Error getting effective level of logger '%s'\n", name);
-      return -1;
-    }
+
+  size_t hash_map_size;
+  rcutils_ret_t hash_map_ret = rcutils_hash_map_get_size(
+    &g_rcutils_logging_severities_map, &hash_map_size);
+  if (hash_map_ret != RCUTILS_RET_OK) {
+    RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting severity level for logger named '%s': %s",
+      name, rcutils_get_error_string().str);
+    return -1;
+  }
+
+  if (hash_map_size == 0) {
+    // The reason we do this extra check here is a bit tricky.  The call to get_severity_level()
+    // below calls rcutils_hash_map_get(), which already has an optimization to return early if
+    // there are no entries in the map.  However, the rest of this function doesn't know how to
+    // disambiguate between not finding this particular name in the map and there being no items in
+    // the map.  In the former case, we need to do the additional work of looking through the
+    // hierarchy of loggers, while in the latter case we can short-circuit.  To keep this function
+    // fast in the latter case, we do the disambiguation here.
+    return g_rcutils_logging_default_logger_level;
+  }
+
+  // Start by trying to find the exact name.
+  int severity;
+  rcutils_ret_t ret = get_severity_level(name, &severity);
+  if (ret == RCUTILS_RET_OK) {
     if (severity != RCUTILS_LOG_SEVERITY_UNSET) {
       return severity;
     }
+    // If the severity is UNSET, then we go through the slow path and try to find a parent logger
+    // that is set.  Failing that, we fall back to the default logger level.
+  } else if (ret != RCUTILS_RET_NOT_FOUND) {
+    // The error message was already set by get_severity_level
+    return -1;
+  }
+
+  // Since we didn't find the name in the fast path, fall back to the slow path where we break the
+  // string into substrings based on dots and look for any part that matches.
```

Comment:
if we want to further optimize this, one thing we could do is to not only store the severity/created_by_user flag of each logger, but also keep a list of the children loggers.

When we create a logger, we also create the parent (and grandparents) if needed, and the new logger is added to the children list.
When the severity of a logger is changed, the severity of the children is also checked and set.
(in the children list, we don't need to keep all children only the ones that inheriting the min severity)

No need to do that in this PR, but I think something like that would work.
Basically, we're having an object for each logger with all the required info, and using a global map to match the objects to names.

Reply:
Yeah, I thought about this approach.

As you say, building a tree like this could significantly speed up the setting of levels.  But I worry that the cost would be slower lookup of levels, which is the thing we really want to optimize here.

Honestly, if we are going to optimize further, I would probably concentrate on your earlier idea of switching to a logger object rather than a string.  In my mind that is likely to be the fastest for both getting and setting.  All of that said, I'm not planning on doing more optimizations here in the near future.

Reply:
> But I worry that the cost would be slower lookup of levels, which is the thing we really want to optimize here.

How the severity level is looked up doesn't change, it's done in the same way.
The tree is used when setting a severity level, instead of iterating over all loggers in the map.
The tree is constructed when a logger is created for the first time(*), by checking all possible parents and adding the new child to the list (if the parent doesn't exist, it's added to the map as well).

Though, the most important thing is the time it takes to query a severity level, and this PR already improves that as much as it seems possible (without breaking API).
So maybe my suggestion will make the code more complicated and improve something that it's not that important.

(*) I think loggers aren't actually "created", I mean when they're queried or their severity level is set for the first time.

> Honestly, if we are going to optimize further, I would probably concentrate on your earlier idea of switching to a logger object rather than a string. In my mind that is likely to be the fastest for both getting and setting.

Agreed!
I was suggesting ideas if we still want to optimize without breaking API.

> All of that said, I'm not planning on doing more optimizations here in the near future.

Yes, that's not needed right now. This is great work!!!